### PR TITLE
spack chain

### PR DESCRIPTION
**Pull request by scheibelp**
@gartung @paul-chelarescu @amundson

See also: https://github.com/spack/spack/pull/8545/

This attempts to connect a local Spack instance to another "upstream" instance, exactly like #8545 (and uses many ideas from it). To use this, see https://github.com/spack/spack/pull/8772#issuecomment-406764998

The primary goal of this PR, quoting from https://github.com/spack/spack/pull/8772#issuecomment-419378823:

> Our use-case for this would be having a "system" instance of Spack maintained by our team that could be "extended" by our users. We would like the "user" instance of Spack to use installed packages in the "system" instance whenever possible.

The implementation differs from #8545: it tries to interact *only* with the upstream database, and otherwise avoids parsing upstream config files with the local Spack instance. For example it does not maintain a `parent_layout` analogous to `parent_db`: it assigns Spec prefixes based on the paths stored in install records and avoids all interactions with the directory layouts of upstream Spack instances.

Similarly, this PR adds a capability for upstream Spack instances to generate a module index file that associates installed packages with their generated module files. This is used for `spack module tcl find` etc. to locate module files. I think this approach will be more resilient to version drift between a local/upstream spack instance. Currently, the module file index is implicitly generated when doing e.g. `spack module tcl refresh`.

(EDIT 11/21: originally the plan to handle module files for upstream packages was to invoke the upstream Spack binary to locate them but that ended up being problematic outside of environment setup scripts. The module indices now correctly locate upstream module files for Spack commands, but do not currently handle autoloading module file generation schemes)

This PR also tries to be careful about not writing any files to upstream database, including:

* Don't acquire locks on upstream DBs
* Don't attempt to reindex upstream DBs if an issue is detected

TODOs (updated 9/7):

- [x] (new: 11/27) update `spack uninstall -a` to uninstall all specs from the local Spack instance (but not attempt to uninstall any upstream Spack installs)
- [x] (pending review) (new: 8/24) update the `csh` environment script to retrieve modules from upstream Spack instances in the same manner `setup-env.sh`
- [x] (new: 9/7) add a test to check that in the simple case where a Spack instance depends on a single upstream instance and a dependency is uninstalled from the upstream instance, that this is reported (e.g. when using `spack find`)
- [x] Raise an exception if an upstream DB attempts to acquire a lock
- [x] Raise an exception if local and upstream DB versions don't match
- [x] Handle loading of modules for packages that are installed upstream
- [ ] (not essential) (new: 8/23) ~update `spack reindex` to set `path` on `InstallRecord`s so that upstream-installed externals can be used for older Spack instances~ (UPDATE) there is not actually sufficient information available for the downstream instance to handle this: Without extra information added to `InstallRecord`, the downstream Spack instance would need to examine the configuration files of the upstream instance to find the externals (which this PR avoids). This PR adds the external path to the `InstallRecord`, so the external package is available; external packages "look the same as" Spack-installed packages from the perspective of a downstream Spack instance. 
- [x] (handled 11/26) Determine whether to support the following case: given databases X, Y, and Z, where X has Y and Z as an upstream, allow Y to use Z as an upstream (if that's desired, it isn't currently supported)
- [ ] (in a future PR) (new: 11/21) handle autoloading module generation schemes where dependencies are associated with upstream Spack instances

### CODE REVIEW

**Comment by scheibelp** [comment]
I won't get a chance to make additional edits to this until 7/30.

**Comment by scheibelp** [comment]
I thought I'd add an example of how you edit config.yaml to include upstream instances but that interferes with doc generation and testing so I'll add an example here:

```
config:
  # For now, you specify the path up to where the DB is stored (everything
  # before .spack-db)
  upstream_spack_installations:
    - /path/to/spack/opt/spack
```

**Comment by paul-chelarescu** [comment]
Hello, I've tested this feature with two installations of spack and two packages, installation 1 with package A and installation 2 with package B -> A and I can confirm that in the second installation spack will only build package B. However I have a case where I point spack to a directory with many pre-built packages and then try to build package C which depends on all of those. Spack will proceed to install every dependency of C again. I am not sure why I get this in the second scenario and I wanted to point it out and that there may be something that is still missing from this feature.

**Comment by scheibelp** [comment]
Hi @paul-chelarescu, as a sanity check: did you confirm that `C` depended on the exact same instance of `B` and `A` (i.e. that the dependency hashes were the same)? I think `spack spec` gives you an option to output the DAG hash (apologies but I'm not in a spot to run `spack spec --help` to make sure of that right now).

If the dependency hashes do match (i.e. they exist in the upstream DB and aren't being used) would it be possible to attach the package.py files for `A/B/C`? I've had luck for at least one case of `A->B->C`, namely `sqlite`, where all of it's dependencies were installed upstream (so perhaps it is clear how your case differs from `sqlite` and you can just mention that vs. including the package.py files).

I'll have a chance to experiment with this on 7/30

**Comment by amundson** [comment]
How hard is it to set up a real test suite implementation of something like this? Obviously, there is a fairly complicated setup required. I haven’t looked into the detail of the test suite at all. I would have created tests like this in my original pull request if I had known how to do it and/or found the time to figure it out.

—Jim
------------------------
James Amundson
Head, Scientific Software Infrastructure Department
Scientific Computing Division, Fermilab
+1 (630) 840-2430



On Jul 25, 2018, at 10:38 PM, scheibelp <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @paul-chelarescu<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_paul-2Dchelarescu&d=DwMFaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=CHBGEtgcObTpDl5qHRk9ayaogsrDBgoaB4wksjoxIZ0&m=eiVF9DLNGWhzCriWTcs7rSvlcfzvTkPB7HBQTO3XEK8&s=-5lGrxzLrS5LDa636Vm7nY3HOs-OrtDvSe_VjFhe_mg&e=>, as a sanity check: did you confirm that C depended on the exact same instance of B and A (i.e. that the dependency hashes were the same)? I think spack spec gives you an option to output the DAG hash (apologies but I'm not in a spot to run spack spec --help to make sure of that right now).

If the dependency hashes do match (i.e. they exist in the upstream DB and aren't being used) would it be possible to attach the package.py files for A/B/C? I've had luck for at least one case of A->B->C, namely sqlite, where all of it's dependencies were installed upstream (so perhaps it is clear how your case differs from sqlite and you can just mention that vs. including the package.py files).

I'll have a chance to experiment with this on 7/30

**Comment by scheibelp** [comment]
> How hard is it to set up a real test suite implementation of something like this?

I anticipate it will be somewhat challenging but I should be able to sort it out by the end of this week. I think either mock packages provided for Spack testing (which makes it easy to create dependency DAGs) or the mock testing repository (which includes packages which actually install) will be useful. More details on that soon.

**Comment by paul-chelarescu** [comment]
@scheibelp I think I have a better clue why C was installing every dependency again, the .spack-db in the upstream was created using an older version of spack (tags/v0.11.0). After your last commit which looks at the spec hash, spack will correctly link against a few of the packages there, but there are some which will fail because of the differences in the database versions. I will try to make another upstream installation with a newer version of spack, but this points out at least that this method is not backwards compatible with older versions of spack installations.

**Comment by scheibelp** [comment]
I've now added tests to check the following:

* `Spec.concretize` appropriately sets `_installed_upstream` on packages (when they are in an upstream DB)
* Packages that are registered in an upstream database are not installed locally
* Upstream packages have the expected prefix (e.g. vs. using the directory layout of the local Spack install)
* Upstream packages are not added to the local DB

I'm going to update the PR description with remaining issues.

> this points out at least that this method is not backwards compatible with older versions of spack installations.

Thanks for testing that! I have not checked what happens when using a different DB version. Minimally I should add an explicit check to assert that the upstream DB is the same version (vs. failing silently).

**Comment by paul-chelarescu** [comment]
@scheibelp I have generated another spack installation in my upstream using a newer version of spack (commit c79cd5fe51352eba3a2855a60ce90421701ed3bc) and ran this feature locally pointing at it. Every package is correctly flagged as being installed upstream, except for one of them:
```
==> Installing gaudi
==> Error: RuntimeError: Unable to locate python command in None/bin

/build/paul/spack/var/spack/repos/builtin/packages/python/package.py:419, in command:
        416            command = 'python'
        417        else:
        418            msg = 'Unable to locate {0} command in {1}'
  >>    419            raise RuntimeError(msg.format(self.name, self.prefix.bin))
        420
        421        # The python command may be a symlink if it was installed
        422        # with Homebrew. Since some packages try to determine the
```

I am trying to investigate the differences between the local installation of this package, which I have been able to link against, and this one above installed upstream which is failing. There are differences between the two index.json databases but it is not clear to me why and what difference exactly is causing this instance to pick up a `None` object.

**Comment by scheibelp** [comment]
> The reason spack cannot find the installation directory of python is that python is externally installed in this case, and the prefix path appears as None.

@paul-chelarescu Thanks for testing this out! I have updated the PR to support this (and added a test for it)

**Comment by scheibelp** [comment]
The errors I'm getting here (specifically with the python 2.6 unit tests) are strange and I cannot replicate them in a debug build, so I'm going to try closing/reopening.

**Comment by scheibelp** [comment]
@paul-chelarescu found issues with creating views using upstream Spack instances and also upstream externals:

* It should now be possible to create views using upstream spack instances
* Using upstream externals is now only supported if those externals were installed using this PR: otherwise the database record does not refer to the external path (this was not originally stored in the database record, and I want to avoid parsing upstream configs). My current thought is that `spack reindex` could rewrite the records, but FYI at the moment reindexing is not likely to fix the problem.
* (implementation detail) I have refactored how upstream installations are detected to make them more robust: originally they depended on `Spec.concretize` to pass over the DAG and mark packages as having been installed upstream. Now there is a `Package.installed_upstream` property which queries the database itself. There are cases where Spack ends up with concrete specs that don't involve concretization (e.g. via `Database.query`)

**Comment by scheibelp** [comment]
>> databases X, Y, and Z, where X has Y and Z as an upstream, allow Y to use Z as an upstream

> what are the main points to address on this feature before it can be merged into develop?

Overall:

1. I need to update the `config.yaml` schema: it needs to be a ordered (part-way through this PR I changed to a dictionary to support module file retrieval) unless the YAML retains the ordering (i.e. in something more-capable than a Python `dict`)
1. I would need to update `store.retrieve_upstream_dbs` - the existing TODO there explains it

I think that's it. I don't consider this feature essential to merge (although I'm curious if you would need it, but I do want to handle (1) before I merge.

**Comment by scheibelp** [comment]
@tgamblin I have removed the WIP tag and now consider this ready for review (FYI, there is one TODO item that I would like to address after a preliminary review).

> I need to update the config.yaml schema: it needs to be a ordered (part-way through this PR I changed to a dictionary to support module file retrieval) unless the YAML retains the ordering (i.e. in something more-capable than a Python dict)

This is now done, and I realized it was important in general, not just to support the case of an upstream DB using its own upstream DB. https://github.com/spack/spack/pull/8772#issuecomment-406764998 has been updated with the appropriate schema.

Regarding support for reindex setting `path` on `InstallRecords`, I realized that `spack reindex` does not have sufficient information to do this properly. That being said this should not be a major issue since external packages were not added to the database until https://github.com/spack/spack/pull/8038 (merged in July).

**Comment by scheibelp** [comment]
> We would like to have this feature merged as a priority and I am happy to provide any essential feedback for it.

I should clarify: I am also prioritizing this feature as a whole. I was just unsure whether I should worry about supporting the case of an upstream DB using its own upstream DB. For now I've decided that can be handled in another PR.

I checked with @paul-chelarescu and that seems to work for him. @healther did you have a different opinion on that?

**Comment by pramodk** [comment]
I am interested in this feature! if I could help with any testing, let me know!

**Comment by healther** [comment]
@scheibelp For our use case a single master->slave relationship would be sufficient. I might be able to test this beginning of October, but currently this is beyond my time budget, sorry.

File: share/spack/setup-env.sh [hunk 1]
```diff
@@ -255,8 +260,15 @@ fi;
 #
 # set module system roots
 #
-_spack_pathadd DK_NODE    "${_sp_dotkit_root%/}/$_sp_sys_type"
-_spack_pathadd MODULEPATH "${_sp_tcl_root%/}/$_sp_sys_type"
+IFS=':' read -ra tcl_roots <<< "$_sp_tcl_roots"
+for tcl_root in "${tcl_roots[@]}"; do
+    _spack_pathadd MODULE_PATH "${tcl_root}/$_sp_sys_type"
```
Comment (luigi-calori) [inline share/spack/setup-env.sh]: I' ve tested this PR with boostrap for environment modules and found that MODULEPATH env var is missing.
I see that you have substituted with MODULE_PATH.
Is this intended or just a typo?
In case of typo, do you mind to fix so to ease testing?

**Comment by luigi-calori** [comment]
I suspect there is a typo in share/spack/setup-env.sh

module env var MODULEPATH has been misspelled as MODULE_PATH

**Comment by scheibelp** [comment]
> module env var MODULEPATH has been misspelled as MODULE_PATH

@luigi-calori Thanks for that! It is now fixed.

**Comment by healther** [comment]
> However, is it currently possible to find packages with broken dependencies in Spack?

There is `spack find -M` "show missing dependencies as well as installed specs". 

From a practical point of view we just enforce a "don't uninstall" policy on the "system" spack, which really only is a problem if you somehow end up changing something in a `package.py` that changes the install but not the hash. But at the same time this really really shouldn't happen. Do you have a concrete use case for uninstalling packages in the system spack?

**Comment by scheibelp** [comment]
> Our use-case for this would be having a "system" instance of Spack maintained by our team that could be "extended" by our users. We would like the "user" instance of Spack to use installed packages in the "system" instance whenever possible.
> As far as I can tell, your PR should cover for that use-case. Am I correct about this?

Yes, that is the number one use case this targets, in fact I'm going to add that quote to the PR description

> One thing that could happen is that a package is uninstalled from the "system" instance of Spack while having dependent packages in some "user" instances. Obviously it is not possible to prevent that and anyway I think it is fine as it is no different from removing a package that might still be used by some users. However, is it currently possible to find packages with broken dependencies in Spack?

Nothing in this PR is known to interfere with the detection of "dangling" dependencies. That being said I should add a test for it.

It comes to mind though that if you have two upstream databases, and a package is uninstalled in one but remains in the other, that Spack could pull in the wrong instance of the package. In general, I want to punt most issues related to uninstalling from upstream to after this PR. However I agree that for the simple case of a single upstream we should support detecting issues.

**Comment by scheibelp** [comment]
Tagging this WIP while I sync with develop and test it. When that's ready (anticipated by the end of Monday 11/19) I'll remove that and it'll be ready for a review.

**Comment by pramodk** [comment]
@scheibelp  : I would like to test this for the deployment related discussion we had during SC. I would like to setup chain as: 

```
- compilers
       |__system-tools
                 |__serial-libraries
                              |__parallel-libraries
                                             |__user-applications
```
where,
* compilers is top level spack
* system-tools chains/inherit compiler spack
* serial-libraries chains/inherits system-tools as well as compiler
* and so on...

Our intention to setup above chain is that the top level spack (i.e. `compilers`) changes very slow and lowest level spack (`our-appliations`) moves very fast (from developers). If we need to re-deploy new release, we would like to re-use most of the packages from top levels and install `user-applications` often.

To test this use case, is this PR ready? Let me know and I can help to test.

One question : can I `re-install / overwrite` spec in the lower level? e.g. `zlib%gcc` present in `serial-libraries` but I would like to reinstall it in `user-applications`? (using `--overwrite` or some other means). (this is strictly not needed but I am curious)

**Comment by matz-e** [comment]
More things me and @pramodk are wondering about:

* does the chain recurse? I.e., we have `tools` linked to `compilers`, and a top-level `applications` linked to `tools`. When I use the Spack installation of `applications` to install software, will it be aware of `compilers` or do I have to add that chain explicitly?
* if we use a python from an upstream Spack, what happens with `spack activate`? Can we block accidental activation in an upstream chain?
  * in this case, we would want to re-install the python locally and then activate said package. Will this workflow be supported?

**Comment by scheibelp** [comment]
@pramodk 

> To test [chaining more than 2 spack instances] use case, is this PR ready?

Yes.

> can I re-install / overwrite spec in the lower level?

Not right now.

> can I add spec from top level spack as an external package with path specification in packages.yaml? e.g. I have installed ‘zlib%gcc’ in ‘serial-libraries’. I want to add that as ‘zlib%intel’ in packages.yaml of ‘parallel-libraries’ to avoid multiple installations.

I would rephrase this as: can I use Spack-installed packages from upstream instances as externals?

Note that you can use packages installed by other instances as externals. You currently cannot use packages installed in the same instance as an external. So the nuance here is whether you can add external entries for packages in Spack instances that are marked as upstream.
 
I did not explicitly set out to support this in this PR but it should work. Is it necessary for you?

@matz-e

> When I use the Spack installation of `applications` to install software, will it be aware of `compilers` or do I have to add that chain explicitly?

They must be added explicitly. If your chain consists of more than 1 upstream database, each database must list out all upstream databases explicitly, and in a consistent order. For example chaining instances `X -> Y -> Z`, `X` must mention `Y` and `Z` in its configuration, and `Y` must mention `Z` in its configuration.

> if we use a python from an upstream Spack, what happens with spack activate?

Currently it would try to write to the upstream Spack's python. A check should be added in Spack to prevent this. Originally I felt that user permissions would be sufficient to prevent this.

I don't think activation could be supported for chained Spack instances. I think views would likely be most appropriate (and should work without modification).

**Comment by scheibelp** [comment]
@becker33 this is now ready for review

**Comment by scheibelp** [comment]
Regarding https://github.com/spack/spack/pull/8772#issuecomment-442040262, it looks like you are referring to the base Spack directory vs. the `opt/spack/` subdirectory underneath the base Spack directory - the latter is required for Spack chain.

Regarding `spack uninstall -a`: it turns out I need to update that to just uninstall specs from the local Spack instance (i.e. I overlooked this). I should be able to handle that this evening.

**Comment by scheibelp** [comment]
> Regarding `spack uninstall -a`: it turns out I need to update that to just uninstall specs from the local Spack instance (i.e. I overlooked this). I should be able to handle that this evening.

This is now handled

File: share/spack/setup-env.sh [hunk 2]
```diff
@@ -273,8 +273,15 @@ fi;
 #
 # set module system roots
 #
-_spack_pathadd DK_NODE    "${_sp_dotkit_root%/}/$_sp_sys_type"
-_spack_pathadd MODULEPATH "${_sp_tcl_root%/}/$_sp_sys_type"
+IFS=':' read -ra tcl_roots <<< "$_sp_tcl_roots"
```
Comment (matz-e) [inline share/spack/setup-env.sh]: `read -a` is a bashism that will break with ZSH (equivalent here: `read -A`, pretty annoying…).

**Comment by pramodk** [comment]
>> can I re-install / overwrite spec in the lower level?

> Not right now.

@scheibelp  : We are testing this further and I think above would useful feature. For example, today we setup central instance providing compilers and some basic serial libraries (one of the library/llvm has python dependency and hence python-3.6.5 got installed). The application team was setting their child instance with python and py-* packages. Python was already in the parent. Py-* got installed but we can't activate the python packages. If there is some way to `blacklist` parent spack python, it would be convenient.

We have also seen below warning but have to find out what exactly causing this :

```
$ spack install -y py-mpi4py@3.0.0 %gcc@6.4.0 hdf5+mpi@1.10.3 %gcc@6.4.0 ....
...
==> Warning: Missing dependency not in database: libxslt/5mlam6x needs libxml2-m7eonfp
==> Warning: Missing dependency not in database: libxslt/5mlam6x needs zlib-w43e56t
==> Warning: Missing dependency not in database: libxslt/5mlam6x needs xz-pjh5xrp
```

**Comment by matz-e** [comment]
> > We have also seen below warning but have to find out what exactly causing this :
> > `$ spack install -y py-mpi4py@3.0.0 %gcc@6.4.0 hdf5+mpi@1.10.3 %gcc@6.4.0 ....`
> > `...`
> > `==> Warning: Missing dependency not in database: libxslt/5mlam6x needs libxml2-m7eonfp`
> > `==> Warning: Missing dependency not in database: libxslt/5mlam6x needs zlib-w43e56t`
> > `==> Warning: Missing dependency not in database: libxslt/5mlam6x needs xz-pjh5xrp`
> 
> One common cause of this for chained spack instances is if the downstream Spack instance isn't pointing at the upstream Spack instance.
> 
> Another possibility is if you have uninstalled packages from the upstream instance.
> 
> Do either of those apply for your case?

I think we had different understandings of the order in the configuration file. Local tests show that we had set up a chain (`X` the most upstream) `X → Y → Z` to have the following configuration:
```yaml
upstream_spack_installations:
- stuff_for_X
- stuff_for_Y
```
If we reverse the order, error messages like the ones you refer to seem to disappear. We'll test with the full deployment, still.

**Comment by matz-e** [comment]
Could we block `buildcache` from creating archives for upstream packages? In our experiments, that has given us some trouble with convoluted relative paths (`../../../my/upstream/chain/installation/root`).

File: lib/spack/spack/directory_layout.py [hunk 3]
```diff
@@ -190,9 +198,6 @@ def hidden_file_paths(self):
     def relative_path_for_spec(self, spec):
         _check_concrete(spec)
 
-        if spec.external:
```
Comment (becker33) [inline lib/spack/spack/directory_layout.py]: Why do we delete these lines?

File: lib/spack/spack/directory_layout.py [hunk 4]
```diff
@@ -74,6 +75,13 @@ def path_for_spec(self, spec):
         """Return absolute path from the root to a directory for the spec."""
         _check_concrete(spec)
 
+        if spec.external:
+            return spec.external_path
+        if self.check_upstream and spec.package.installed_upstream:
+            raise SpackError(
```
Comment (becker33) [inline lib/spack/spack/directory_layout.py]: Why not just return the upstream path here?

File: lib/spack/spack/spec.py [hunk 5]
```diff
@@ -1384,7 +1384,12 @@ def cshort_spec(self):
     @property
     def prefix(self):
         if self._prefix is None:
-            self.prefix = spack.store.layout.path_for_spec(self)
+            upstream, record = spack.store.db.query_by_spec_hash(
+                self.dag_hash())
+            if upstream:
```
Comment (becker33) [inline lib/spack/spack/spec.py]: I think this check should be on record, not upstream. Whether we found the spec locally or upstream, if we found a record we should use the path from that record, and only compute the path when the spec isn't found.

This will be important for things like if a user changes their install layout. They should still be able to see the packages they installed before the change, but they won't if we recompute the prefix when we could've gotten it from the database.

File: share/spack/setup-env.csh [hunk 6]
```diff
@@ -24,6 +24,7 @@ if ($?SPACK_ROOT) then
     _spack_pathadd PATH "$SPACK_ROOT/bin"
     eval `spack --print-shell-vars csh`
 
+    # TODO: this needs to be updated the same way as setup-env.sh
```
Comment (becker33) [inline share/spack/setup-env.csh]: Still have a TODO left here.

File: lib/spack/spack/store.py [hunk 7]
```diff
@@ -84,3 +85,28 @@ def _store():
...
+
+    install_roots = []
+    for install_properties in other_spack_instances:
+        install_roots.append(install_properties['spack-install-prefix'])
+
+    return _construct_upstream_dbs_from_install_roots(install_roots)
+
+
+def _construct_upstream_dbs_from_install_roots(
+        install_roots, _test=False):
```
Comment (becker33) [inline lib/spack/spack/store.py]: I don’t like that we have separate code paths for testing… I’d rather somehow capture the warning output and fail the test based on that, but separate code paths scare me from a maintenance perspective and an accuracy of testing perspective.

File: lib/spack/spack/directory_layout.py (hunk 3 above)
Comment (scheibelp) [reply lib/spack/spack/directory_layout.py]: I moved this handling into `path_for_spec`

File: lib/spack/spack/store.py (hunk 7 above)
Comment (scheibelp) [reply lib/spack/spack/store.py]: I may be missing something: this function is tested and also used by `retrieve_upstream_dbs`, so I don't regard the code paths as separate. This function isolates the most-easily-tested logic. If you are referring to the `_test` variable, the only effect it has right now is to throw an error in place of printing a warning, so that the testing framework can confirm the behavior. Was that still problematic or was there another concern I missed?

File: lib/spack/spack/directory_layout.py (hunk 4 above)
Comment (scheibelp) [reply lib/spack/spack/directory_layout.py]: My philosophy is that this function should not be called for upstream-installed packages and doing so implies there is an error in the Spack core.

File: lib/spack/spack/spec.py (hunk 5 above)
Comment (scheibelp) [reply lib/spack/spack/spec.py]: If I understand correctly, you are describing the case that a user installs some packages with Spack, and then alters the `install_path_scheme`.

I agree that what you suggest is necessary to support changing the install layout, but I don't know if it is sufficient - are you more certain about that? I'd prefer to handle that in a different PR.

Additionally, this would likely require some additional care. For one, `if not record.installed` then the path should also be taken from `spack.store.layout`.

**Comment by scheibelp** [comment]
@becker33 

> The downstream version reports both in spack find (problem 1), and reports them both as being installed locally via spack find -p (problem 2). The query function should filter out upstream packages that match the hashes of local packages.

Both of these are now addressed.

https://github.com/spack/spack/pull/8772#discussion_r239559866 is also addressed. I have added responses to other review comments for clarification.

**Comment by scheibelp** [comment]
The config yaml format has been updated: full details are at https://github.com/spack/spack/pull/8772#issuecomment-406764998 (along with a mention of how this is different from before). In short to use this you'll need to move the configuration into an `upstreams.yaml` file and tweak it a bit.

**Comment by JavierCVilla** [comment]
@scheibelp What's the current status of this feature?

File: lib/spack/spack/test/database.py [hunk 8]
```diff
@@ -17,13 +17,171 @@
...
+        prepared_db._write_to_file(db_file)
+
+    default = ('build', 'link')
+    x = MockPackage('x', [], [])
+    z = MockPackage('z', [], [])
+    y = MockPackage('y', [z], [default])
+    w = MockPackage('w', [x, y], [default, default])
+    mock_repo = MockPackageMultiRepo([w, x, y, z])
+
+    mock_layout = MockLayout('/a/')
```
Comment (becker33) [inline lib/spack/spack/test/database.py]: I think this should use tmpdir for the layout root. I don't think this should write to the layout, but if something goes wrong we don't want to introduce side-effects.

File: lib/spack/spack/test/database.py [hunk 9]
```diff
@@ -17,13 +17,171 @@
...
+        spec.concretize()
+
+        for dep in spec.traverse(root=False):
+            prepared_db.add(dep, mock_layout)
+
+        downstream_db_root = str(
+            tmpdir_factory.mktemp('mock_downstream_db_root'))
+        downstream_db = spack.database.Database(
+            downstream_db_root, upstream_dbs=[prepared_db])
+        downstream_layout = MockLayout('/b/')
```
Comment (becker33) [inline lib/spack/spack/test/database.py]: Similarly, use a tmpdir for the layout root.

File: lib/spack/spack/test/database.py (hunk 8 above)
Comment (becker33) [inline lib/spack/spack/test/database.py]: Can you rename the upstream pieces that you called "mock_"- or "prepared_"- to "upstream_"-? I think that will make maintenance easier.

File: lib/spack/spack/test/database.py [hunk 10]
```diff
@@ -17,13 +17,171 @@
...
+    # Generate initial DB file to avoid reindex
+    with open(prepared_db._index_path, 'w') as db_file:
+        prepared_db._write_to_file(db_file)
+
+    default = ('build', 'link')
+    z = MockPackage('z', [], [])
+    y = MockPackage('y', [z], [default])
+    mock_repo = MockPackageMultiRepo([y, z])
+
+    mock_layout = MockLayout('/a/')
```
Comment (becker33) [inline lib/spack/spack/test/database.py]: Same here

File: lib/spack/spack/test/database.py [hunk 11]
```diff
@@ -17,13 +17,171 @@
...
+            downstream_db_root, upstream_dbs=[prepared_db])
+        new_downstream._fail_when_missing_deps = True
+        with pytest.raises(spack.database.MissingDependenciesError):
+            new_downstream._read()
+
+
+@pytest.mark.usefixtures('config')
+def test_recursive_upstream_dbs(tmpdir_factory, test_store):
+    roots = [str(tmpdir_factory.mktemp(x)) for x in ['a', 'b', 'c']]
+    layouts = [MockLayout(x) for x in ['/ra/', '/rb/', '/rc/']]
```
Comment (becker33) [inline lib/spack/spack/test/database.py]: Same here

File: lib/spack/spack/test/install.py [hunk 12]
```diff
@@ -125,6 +125,75 @@ def test_installed_dependency_request_conflicts(
...
 
+def test_installed_upstream_external(
+        tmpdir_factory, install_mockery, mock_fetch):
+    """Check that when a dependency package is recorded as installed in
+       an upstream database that it is not reinstalled.
+    """
+    mock_db_root = str(tmpdir_factory.mktemp('mock_db_root'))
+    prepared_db = spack.database.Database(mock_db_root)
+
+    mock_layout = spack.test.database.MockLayout('/a/')
```
Comment (becker33) [inline lib/spack/spack/test/install.py]: Same here

File: share/spack/setup-env.sh (hunk 2 above)
Comment (becker33) [reply share/spack/setup-env.sh]: That's new to me, good catch! We should clean that up

File: lib/spack/spack/test/install.py [hunk 13]
```diff
@@ -125,6 +125,75 @@ def test_installed_dependency_request_conflicts(
...
+        assert new_dependency.prefix == mock_layout.path_for_spec(dependency)
+
+        dependent.package.do_install()
+
+        assert not os.path.exists(new_dependency.prefix)
+        assert os.path.exists(dependent.prefix)
+    finally:
+        spack.store.db = original_db
+
+
```
Comment (becker33) [inline lib/spack/spack/test/install.py]: Can you add a test for the errors I brought up in my previous review? Namely that database queries were returning duplicate packages in the case the package was installed upstream after it was installed downstream.

File: lib/spack/spack/store.py (hunk 7 above)
Comment (becker33) [reply lib/spack/spack/store.py]: This still concerns me, but it doesn't need to change as part of this PR.

**Comment by becker33** [comment]
@JavierCVilla it was sitting on my plate to review until this afternoon.

File: share/spack/setup-env.sh (hunk 2 above)
Comment (matz-e) [reply share/spack/setup-env.sh]: I've attempted a fix [here](https://github.com/scheibelp/spack/pull/2).

**Comment by scheibelp** [comment]
FYI I have addressed or responded to all review comments except https://github.com/spack/spack/pull/8772#discussion_r261443812 (adding a test for when a package is installed both upstream and downstream). I will update this when I add that.

**Comment by scheibelp** [comment]
Whoops I accidentally closed it when I meant to just comment

File: lib/spack/spack/main.py [hunk 14]
```diff
@@ -571,9 +571,28 @@ def shell_set(var, value):
...
+        'tcl': list(),
+        'dotkit': list(),
+        'lmod': list()
+    }
     for name, path in module_roots.items():
         path = spack.util.path.canonicalize_path(path)
-        shell_set('_sp_%s_root' % name, path)
+        module_to_roots[name].append(path)
+
+    other_spack_instances = spack.config.get(
```
[the remark this answers is not in the document]
Comment (becker33) [reply lib/spack/spack/main.py]: TIL. +1

File: lib/spack/spack/database.py [hunk 15]
```diff
@@ -657,6 +729,11 @@ def _add(
             raise NonConcreteSpecAddError(
                 "Specs added to DB must be concrete.")
 
+        key = spec.dag_hash()
+        upstream, record = self.query_by_spec_hash(key)
+        if upstream:
```
[the remark this answers is not in the document]
Comment (becker33) [reply lib/spack/spack/database.py]: > All upstream DBs are read once, in `store.py`, so the case you mention would not occur.

As long as we don't re-read upstream DBs on query, that should be fine.

File: lib/spack/spack/package.py [hunk 16]
```diff
@@ -490,6 +490,15 @@ def __init__(self, spec):
 
         super(PackageBase, self).__init__()
 
+    @property
+    def installed_upstream(self):
+        if not hasattr(self, '_installed_upstream'):
+            upstream, record = spack.store.db.query_by_spec_hash(
+                self.spec.dag_hash())
+            self._installed_upstream = upstream
```
[the remark this answers is not in the document]
Comment (becker33) [reply lib/spack/spack/package.py]: I think the read-once issue we discussed above should take care of this as well. I'll resolve it.

File: lib/spack/spack/directory_layout.py [hunk 17]
```diff
@@ -223,19 +228,24 @@ def spec_file_path(self, spec):
         return os.path.join(self.metadata_path(spec), self.spec_file_name)
 
     def metadata_path(self, spec):
-        return os.path.join(self.path_for_spec(spec), self.metadata_dir)
+        if self.check_upstream and spec.package.installed_upstream:
+            # TODO: This assumes that older spack versions use the same
```
[the remark this answers is not in the document]
Comment (becker33) [reply lib/spack/spack/directory_layout.py]: The metadata dir is configurable on the YamlDirectoryLayout object. I think we should remove that option (no code appears to use it) as part of this PR since we depend on it not being used in the future.

File: share/spack/setup-env.sh (hunk 2 above)
Comment (matz-e) [reply share/spack/setup-env.sh]: @scheibelp couldn't this be changed for ZSH users?

**Comment by scheibelp** [comment]
https://github.com/spack/spack/pull/8772#discussion_r261443812 and https://github.com/spack/spack/pull/8772#discussion_r264864056 are now addressed. Also, note that https://github.com/spack/spack/pull/8772/commits/5c4330b6b4cdbf61d381685a73608ea7926955d4 is addressed by https://github.com/spack/spack/pull/8772/commits/beb9a31e5189d89e845f5724115d863b302da685.
 
I decided as part of this that YamlDirectoryLayout should defer entirely to Spec.prefix when constructing the metadata path, which meant a test in `directory_layout` had to be updated to update `spack.store.layout` (see the `layout_and_dir` fixture update in https://github.com/spack/spack/pull/8772/commits/2e39fdfc780f1f30d09fbf2f6ff7e10e410ff413).

**Comment by KineticTheory** [comment]
I'm really excited to see this merged and I am ready to try it again.  I know the yaml files changed format several times during the development process.  Is there an official documentation location yet? I didn't see anything about `spack chain` at https://spack.readthedocs.io.

**Comment by gartung** [comment]
@chissg @drbenmorgan

**Comment by chissg** [comment]
On 3/27/19 5:06 PM, Patrick Gartung wrote:

@chissg<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_chissg&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=kj1IkzpbbOO6DafO4zbQ7w&m=XhUe9qhSJGagq5lSpAifwwjPA1mweTqGnOEcvvI_E24&s=FZQvCabipDRLAIZQFWUd6tVYa32BoveIqHoOZWDpC-o&e=> @drbenmorgan<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_drbenmorgan&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=kj1IkzpbbOO6DafO4zbQ7w&m=XhUe9qhSJGagq5lSpAifwwjPA1mweTqGnOEcvvI_E24&s=PirrIZU4RvFSeeYCV27VXRNSIc1QMDGfaO9utkwnjgc&e=>

Already merged into https://github.com/FNALssi/spack/commits/fnal-develop.

C.

**Comment by sknigh** [comment]
@scheibelp I want to try this out, is any documentation available?

**Comment by scheibelp** [comment]
@sknigh 

> I want to try this out, is any documentation available?

not in the actual docs, it is all here for now (the PR description and https://github.com/spack/spack/pull/8772#issuecomment-406764998). Although docs are definitely worth adding and I can add a PR for something official in the next few days.
